integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/database/device/DlmsDeviceSteps.java

```diff
@@ -751,20 +813,128 @@ public void registerNewKeys(final Map<String, String> inputSettings) {
+          dbEncryptedSecret.stream()
+              .map(DbEncryptedSecret::getEncodedSecret)
+              .collect(Collectors.toList());
+      assertThat(actualEncodedSecrets)
+          .withFailMessage(
+              "Wrong dbEncryptedSecret for %s with status %s expected %s to be contained in: %s",
+              secretTypeString, secretStatus, dbEncryptedSecretValue, actualEncodedSecrets)
+          .contains(dbEncryptedSecretValue);
+    }
+  }
+
+  @Then(
+      "the encrypted_secret table in the secret management database should contain an EXPIRED and an ACTIVE key for device {string} or exactly the opposite")
+  public void
+      theEncryptedSecretTableInTheSecretManagementDatabaseShouldContainKeysForDeviceOrExactlyTheOpposite(
+          final String deviceIdentification, final Map<String, String> inputSettings) {
+    // Depending on the order in which the two request are processed the results (status of the
+    // secret) are either as specified as the first or second value
+    int countResultKeysRequest1 = 0;
+    int countResultKeysRequest2 = 0;
+
+    for (final String secretTypeString : inputSettings.keySet()) {
+
+      final String[] keyInfo = inputSettings.get(secretTypeString).split(",");
+
+      final String keyName1 = keyInfo[0];
+      final String keyName2 = keyInfo[1];
+
+      final SecretType secretType = this.getSecretTypeByKeyTypeInputName(secretTypeString);
+
+      final List<DbEncryptedSecret> dbEncryptedSecretsRequest1 =
+          this.encryptedSecretRepository.findSecretsWithSpecificEncodedSecret(
+              deviceIdentification,
+              secretType,
+              SecretStatus.EXPIRED,
+              SecurityKey.valueOf(keyName1).getDatabaseKey());
+      dbEncryptedSecretsRequest1.addAll(
+          this.encryptedSecretRepository.findSecretsWithSpecificEncodedSecret(
+              deviceIdentification,
+              secretType,
+              SecretStatus.ACTIVE,
+              SecurityKey.valueOf(keyName2).getDatabaseKey()));
+
+      countResultKeysRequest1 += dbEncryptedSecretsRequest1.size();
+
+      final List<DbEncryptedSecret> dbEncryptedSecretsRequest2 =
+          this.encryptedSecretRepository.findSecretsWithSpecificEncodedSecret(
+              deviceIdentification,
+              secretType,
+              SecretStatus.ACTIVE,
+              SecurityKey.valueOf(keyName1).getDatabaseKey());
+      dbEncryptedSecretsRequest2.addAll(
+          this.encryptedSecretRepository.findSecretsWithSpecificEncodedSecret(
+              deviceIdentification,
+              secretType,
+              SecretStatus.EXPIRED,
+              SecurityKey.valueOf(keyName2).getDatabaseKey()));
+
+      countResultKeysRequest2 += dbEncryptedSecretsRequest2.size();
+    }
+
+    assertThat(
+            (countResultKeysRequest1 == 0 && countResultKeysRequest2 == 4)
+                || (countResultKeysRequest1 == 4 && countResultKeysRequest2 == 0))
+        .isTrue();
+  }
```

Comment:
I'd probably go for something like the following:
```java
  @Then(
      "the encrypted_secret table in the secret management database should contain EXPIRED and one ACTIVE key for device {string} or exactly the opposite")
  public void
      theEncryptedSecretTableInTheSecretManagementDatabaseShouldContainKeysForDeviceOrExactlyTheOpposite(
          final String deviceIdentification, final Map<String, String> inputSettings) {
    // Depending on the order in which the two request are processed the results (status of the
    // secret) are either as specified as the first or second value

    final List<String> activeAuthenticationKeys =
        this.encryptedSecretRepository
            .findSecrets(
                deviceIdentification, SecretType.E_METER_AUTHENTICATION_KEY, SecretStatus.ACTIVE)
            .stream()
            .map(DbEncryptedSecret::getEncodedSecret)
            .collect(Collectors.toList());
    final List<String> expiredAuthenticationKeys =
        this.encryptedSecretRepository
            .findSecrets(
                deviceIdentification, SecretType.E_METER_AUTHENTICATION_KEY, SecretStatus.EXPIRED)
            .stream()
            .map(DbEncryptedSecret::getEncodedSecret)
            .collect(Collectors.toList());
    final List<String> activeEncryptionKeys =
        this.encryptedSecretRepository
            .findSecrets(
                deviceIdentification,
                SecretType.E_METER_ENCRYPTION_KEY_UNICAST,
                SecretStatus.ACTIVE)
            .stream()
            .map(DbEncryptedSecret::getEncodedSecret)
            .collect(Collectors.toList());
    final List<String> expiredEncryptionKeys =
        this.encryptedSecretRepository
            .findSecrets(
                deviceIdentification,
                SecretType.E_METER_ENCRYPTION_KEY_UNICAST,
                SecretStatus.EXPIRED)
            .stream()
            .map(DbEncryptedSecret::getEncodedSecret)
            .collect(Collectors.toList());

    final String[] authenticationKeyNames =
        inputSettings.get(PlatformSmartmeteringKeys.KEY_DEVICE_AUTHENTICATIONKEY).split(",");
    final String[] encryptionKeyNames =
        inputSettings.get(PlatformSmartmeteringKeys.KEY_DEVICE_ENCRYPTIONKEY).split(",");

    final int numberOfRequestsFromWhichKeysAreIncluded = authenticationKeyNames.length;
    int countKeysFromRequestBothActive = 0;
    int countKeysFromRequestBothExpired = 0;

    for (int i = 0; i < numberOfRequestsFromWhichKeysAreIncluded; i++) {
      final String authenticationKeyFromRequest =
          SecurityKey.valueOf(authenticationKeyNames[i]).getDatabaseKey();
      final String encryptionKeyFromRequest =
          SecurityKey.valueOf(encryptionKeyNames[i]).getDatabaseKey();

      if (activeAuthenticationKeys.contains(authenticationKeyFromRequest)
          && activeEncryptionKeys.contains(encryptionKeyFromRequest)) {
        countKeysFromRequestBothActive += 1;
      } else if (expiredAuthenticationKeys.contains(authenticationKeyFromRequest)
          && expiredEncryptionKeys.contains(encryptionKeyFromRequest)) {
        countKeysFromRequestBothExpired += 1;
      } else {
        throw new AssertionError(
            "Keys from the same request ended up not both active or both expired.");
      }
    }

    assertThat(countKeysFromRequestBothActive).isOne();
    assertThat(countKeysFromRequestBothExpired)
        .isEqualTo(numberOfRequestsFromWhichKeysAreIncluded - 1);
  }
```
I have not tested above suggestion, but I think it is a bit more clear as (in my opinion) there is less "figuring out" to do to match calculated numbers from different requests to what is expected.
I think it has the advantage that a method like `findSecretsWithSpecificEncodedSecret` which was only useful for the assertions in the test and not for any production scenario could be avoided altogether.
It could be used with even more requests than two (though this is admittedly not something you would likely want to do).
You can treat this as a (firm) suggestion, opinions on clarity and readability may differ.

Reply:
Nice rework. This check also works for more than two parallel requests so I change the name of the check to
`the encrypted_secret table in the secret management database should contain one or more EXPIRED and just one ACTIVE key for device {string} in correct combination`
Extended the scenario to 3 parallel requests
Removed altogether the catch on more than two requests (as mentioned in comment above)